rollup/ccc/logger.go

```diff
@@ -0,0 +1,859 @@
+package ccc
+
+import (
+	"maps"
+	"math/big"
+	"time"
+
+	"github.com/scroll-tech/go-ethereum/common"
+	"github.com/scroll-tech/go-ethereum/core/types"
+	"github.com/scroll-tech/go-ethereum/core/vm"
+	"github.com/scroll-tech/go-ethereum/crypto"
+	"github.com/scroll-tech/go-ethereum/eth/tracers"
+)
+
+const (
+	sigCountMax        = 127
+	ecAddCountMax      = 50
+	ecMulCountMax      = 50
+	ecPairingCountMax  = 2
+	rowUsageMax        = 1_000_000
+	keccakRounds       = 24
+	keccakRowsPerRound = 12
+	keccakRowsPerChunk = (keccakRounds + 1) * keccakRowsPerRound
+)
+
+var _ vm.EVMLogger = (*Logger)(nil)
+
+// Logger is a tracer that keeps track of resource usages of each subcircuit
+// that Scroll's halo2 based zkEVM has. Some subcircuits are not tracked
+// here for the following reasons.
+//
+// rlp: worker already keeps track of how big a block is and the block size limit
+// it uses is way below what the rlp circuit allows
+// pi: row usage purely depends on the number txns and we already have a limit
+// on how many txns that worker will pack in a block
+// poseidon: not straight forward to track in block building phase. We can do
+// worst case estimation down the line if needed.
+// mpt: not straight forward to track in block building phase. We can do
+// worst case estimation down the line if needed.
+// tx: row usage depends on the length of raw txns and the number of storage
+// slots and/or accounts accessed. With the current gas limit of 10M, it is not possible
+// to overflow the circuit.
+type Logger struct {
+	currentEnv    *vm.EVM
+	isCreate      bool
+	codesAccessed map[common.Hash]bool
+
+	evmUsage       uint64
+	stateUsage     uint64
+	bytecodeUsage  uint64
+	sigCount       uint64
+	ecAddCount     uint64
+	ecMulCount     uint64
+	ecPairingCount uint64
+	copyUsage      uint64
+	sha256Usage    uint64
+	expUsage       uint64
+	modExpUsage    uint64
+	keccakUsage    uint64
+
+	l2TxnsRlpSize uint64
+}
+
+func NewLogger() *Logger {
+	const miscKeccakUsage = 50_000 // heuristically selected safe number to account for Rust side implementation details
+	return &Logger{
+		codesAccessed: make(map[common.Hash]bool),
+		keccakUsage:   miscKeccakUsage,
+	}
+}
+
+// Snapshot creates an independent copy of the logger
+func (l *Logger) Snapshot() *Logger {
+	newL := *l
+	newL.codesAccessed = maps.Clone(newL.codesAccessed)
+	return &newL
```

Comment:
It will get overwritten when the snapshot is used for the first time, but it is safer to set it to nil here. Thanks! https://github.com/scroll-tech/go-ethereum/pull/982/commits/50636953bdcfc2bf0a7db4f648e85299e5242a2c